quantum/mousekey.c

```diff
@@ -17,19 +17,27 @@
 
 #include <stdint.h>
 #include <string.h>
+#include <math.h>
 #include "keycode.h"
 #include "host.h"
 #include "timer.h"
 #include "print.h"
 #include "debug.h"
 #include "mousekey.h"
+#include "lib/lib8tion/lib8tion.h"
 
 static inline int8_t times_inv_sqrt2(int8_t x) {
+#ifdef MOUSEKEY_PRECISE_DIAGONAL_MOVE
+    // 46341/65536 is very close to 1/sqrt(2)
+    // 0.70710754                   0.707106781
+    return round((x < 0 ? -1 : 1) * scale16(x < 0 ? -x : x, 46341));
+#else
     // 181/256 is pretty close to 1/sqrt(2)
     // 0.70703125                 0.707106781
     // 1 too small for x=99 and x=198
     // This ends up being a mult and discard lower 8 bits
     return (x * 181) >> 8;
+#endif
```

Comment:
```suggestion
#include "keycode.h"
#include "host.h"
#include "timer.h"
#include "print.h"
#include "debug.h"
#include "mousekey.h"

static inline int8_t times_inv_sqrt2(int8_t x) {
    // 181/256 is pretty close to 1/sqrt(2)
    // 0.70703125                 0.707106781
    int16_t n = x * 181, d = 256;
    // Compare signs of numerator and denominator
    // for integer division rounding
    if ((n < 0) ^ (d < 0)) {
        // If signs are opposite, subtract half of
        // the denominator before dividing
        return (n - d / 2) / d;
    } else {
        // If signs are the same, add half of
        // the denominator before dividing
        return (n + d / 2 ) / d;
    }
```

Can you test and consider this? My result shows that it matches `round(x * 0.707106781)` and will avoid the use of expensive `float` operations. Its AVR overhead is ~28 bytes and should not need preprocessor gating.

Reply:
Okay. I'll test it.

Shouldn't it be a second option? Or is it consuming less memory enough to substitute for the `scale16()` method?

Reply:
> Okay. I'll test it.
> 
> Shouldn't it be a second option? Or is it consuming less memory enough to substitute for the `scale16()` method?

`scale16()` unsigned output are rounded down. The ternary statement above will round the integer correctly after division.

Reply:
I meant, if your newest method consumes relatively much resources rather than scale16(), users should be able to choose these 2 options.

Reply:
> I meant, if your newest method consumes relatively much resources rather than scale16(), users should be able to choose these 2 options.

The ternary statement uses less space than `scale16()`.

Reply:
But after thinking a bit, I realized that it's too complicated.

Reply:
Why is it complicated? Did you check if the output is the same as `round(x * 0.707106781)`?

Reply:
No no, preparing 2 options seems to be complicated for users.
And I understood the ternary method is precise and needs less small memory.

That's why I will test for replacing `scale16()` with ternary one.

Reply:
I'm out now. So I'll check it later.